### PR TITLE
GraphQL GiftMessageGraphQl add coverage for cart

### LINKED ISSUE #28519: GraphQL Checkout :: Add support for Gift Wrapping / Gift Message during checkout.
As a Magento Merchant,
I would like the ability to offer my customers gift wrapping and gift messaging options during checkout 
so that 
they can send the products purchased from my business as a gift to their friends/families etc. 

 

**Acceptance Criteria**
GraphQL provides all available gift options for a store.  
GraphQL provides the option to select the gift option for each line item 
GraphQL provides the option to add a gift message for the entire cart and for each item in the cart.

**Approved Schema**

- https://github.com/magento/architecture/blob/master/design-documents/graph-ql/coverage/gift-wrapping.md

### PR DESCRIPTION
**Pull request by Usik2203**
### Description (*)

This PR extends the existing GraphQl coverage by adding information about gift message to different types of cart items.

**Please note**, this PR does not cover the entire gift wrapping functionality but resolvers for the gift message. The rest of the git wrapping functionality will be delivered with the others PR
### Note: These types were not extended  because these types are not declared
<pre> 
type GiftCardCartItem {
    gift_message: GiftMessage @doc(description: "The entered gift message for the cart item")
}

type SalesItemInterface {
    gift_wrapping: GiftWrapping @doc(description: "The selected gift wrapping for the order item")
    gift_message: GiftMessage @doc(description: "The entered gift message for the order item")
}

type SalesTotalsInterface {
    gift_options: GiftOptionsPrices @doc(description: "The list of prices for the selected gift options")
}
</pre>

This PR was mentioned for fixing failed static test related with composer.json file
[#253](https://github.com/magento/partners-magento2ee/pull/253)

### Related Pull Requests
#28072
#28105 

### Fixed Issues (if relevant)

1. [GraphQL Checkout :: Add support for Gift Wrapping / Gift Message during checkout.](https://github.com/magento/community-features/issues/256)

### Manual testing scenarios (*)

You can use this query for manual testing. You should take `maskedQuoteId` from `quote_id_mask` table
<pre>
{
    cart(cart_id: "maskedQuoteId") {
           gift_message {
               to
               from
               message
           }
      }
}
</pre>

### CODE REVIEW

File: app/code/Magento/GiftMessageGraphQl/Model/Resolver/GiftMessage.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    )  {
+        if (!isset($value['model'])) {
+            throw new GraphQlInputException(__('"model" value should be specified'));
+        }
+        $cart = $value['model'];
+
+        $giftCartMessage = $this->cartRepository->get($cart->getId());
+
+        return [
+            'to' => $giftCartMessage->getRecipient(),
```
Comment (rogyar) [inline app/code/Magento/GiftMessageGraphQl/Model/Resolver/GiftMessage.php]: The `recipient`, `sender` and `message` may have `null` value. In this case, we need to make sure that we have no exception because all 3 values are required (and have `Sring` type) in the schema. I would recommend using the null coalesce operator and return empty string in case of no value. i.e. 

```php
$giftCartMessage->getSender() ?? ''
```

File: app/code/Magento/GiftMessageGraphQl/composer.json [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  "name": "magento/module-gift-message-graph-ql",
+  "description": "N/A",
+  "type": "magento2-module",
+  "require": {
+    "php": "~7.1.3||~7.2.0||~7.3.0",
+      "magento/framework": "*",
+      "magento/module-backend": "*",
+      "magento/module-catalog": "*",
+      "magento/module-checkout": "*",
+      "magento/module-customer": "*",
```
Comment (rogyar) [inline app/code/Magento/GiftMessageGraphQl/composer.json]: We need to make sure that we need all these dependencies

File: app/code/Magento/GiftMessageGraphQl/Model/Resolver/GiftMessageItem.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+use Magento\Framework\GraphQl\Config\Element\Field;
+use Magento\Framework\GraphQl\Query\Resolver\ContextInterface;
+use Magento\Framework\GraphQl\Query\Resolver\Value;
+use Magento\Framework\GraphQl\Query\ResolverInterface;
+use Magento\Framework\GraphQl\Schema\Type\ResolveInfo;
+
+/**
+ * Class GiftMessageItem
+ */
+class GiftMessageItem implements ResolverInterface
```
Comment (rogyar) [inline app/code/Magento/GiftMessageGraphQl/Model/Resolver/GiftMessageItem.php]: I'm not sure we should extend this PR with more functionality. Let's leave the gift messages for different cart item types in this PR and cover all other functionality (mutations, gift wrapping) in the following PRs

**Review by rogyar — CHANGES_REQUESTED** [review]
Hi @Usik2203. Thank you for your collaboration. Please, check my comments, adjust the code style and you may proceed with the test coverage. 

Thank you!

File: app/code/Magento/GiftMessageGraphQl/Model/Resolver/GiftMessage.php (hunk 1 above)
Comment (Usik2203) [reply app/code/Magento/GiftMessageGraphQl/Model/Resolver/GiftMessage.php]: **Done**

File: app/code/Magento/GiftMessageGraphQl/composer.json (hunk 2 above)
Comment (Usik2203) [reply app/code/Magento/GiftMessageGraphQl/composer.json]: **Done**

File: app/code/Magento/GiftMessageGraphQl/Model/Resolver/GiftMessageItem.php (hunk 3 above)
Comment (Usik2203) [reply app/code/Magento/GiftMessageGraphQl/Model/Resolver/GiftMessageItem.php]: I have removed it from this PR, will do it in other PR

**Comment by Usik2203** [comment]
@rogyar api-functional test was added for this PR. 
Thank you for review.

File: app/code/Magento/GiftMessageGraphQl/README.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# GiftMessageGraphQl
+
+**GiftMessageGraphQl**
```
Comment (rogyar) [inline app/code/Magento/GiftMessageGraphQl/README.md]: Please, provide a short description of the module purpose

File: app/code/Magento/GiftMessageGraphQl/composer.json [hunk 5]
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "magento/module-gift-message-graph-ql",
+  "description": "N/A",
+  "type": "magento2-module",
+  "require": {
```
Comment (rogyar) [inline app/code/Magento/GiftMessageGraphQl/composer.json:5]: For module responsible for handling the GraphQl we need at least the GraphQl module itself. So we need to add it to the requirements. Please, check other GraphQl modules as a reference

**Review by rogyar — CHANGES_REQUESTED** [review]

**Comment by rogyar** [comment]
Hi @Usik2203. Considering the fact that this solution will be tested manually on some stage, it would be very useful to have the corresponding section in the description filled out. It will be enough to post an example of the GraphQL request and some expected result. 

Thank you!

File: app/code/Magento/GiftMessageGraphQl/README.md (hunk 4 above)
Comment (Usik2203) [reply app/code/Magento/GiftMessageGraphQl/README.md]: **Done**

File: app/code/Magento/GiftMessageGraphQl/composer.json (hunk 5 above)
Comment (Usik2203) [reply app/code/Magento/GiftMessageGraphQl/composer.json:5]: Added `"magento/module-graph-ql": "*"` as dependency

**Comment by Usik2203** [comment]
Hi @rogyar . I added example of query for manual testing in **Manual testing scenarios** area.
Thanks

**Comment by rogyar** [comment]
@Usik2203 please, take a look at the failing tests. Thank you

File: app/code/Magento/GiftMessageGraphQl/composer.json [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    "php": "~7.1.3||~7.2.0||~7.3.0",
```
Comment (danielrenaud) [inline app/code/Magento/GiftMessageGraphQl/composer.json]: Can you update php requirements to match recently updated versions for 2.4

File: app/code/Magento/GiftMessageGraphQl/Model/Resolver/Cart/GiftMessage.php [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    ) {
+        if (!isset($value['model'])) {
+            throw new GraphQlInputException(__('"model" value should be specified'));
+        }
+        $cart = $value['model'];
+
+        $giftCartMessage = $this->cartRepository->get($cart->getId());
+
+        if ($giftCartMessage === null) {
+            throw new GraphQlInputException(__("Gift message doesn't exist for current cart"));
```
Comment (danielrenaud) [inline app/code/Magento/GiftMessageGraphQl/Model/Resolver/Cart/GiftMessage.php]: Is it appropriate to throw exception in this case? A cart without a gift message doesn't seem like an exceptional case; probably we can just return something empty.

File: dev/tests/api-functional/testsuite/Magento/GraphQl/GiftMessage/Cart/GiftMessageTest.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+namespace Magento\GraphQl\GiftMessage\Cart;
+
+use Exception;
+use Magento\Framework\Exception\NoSuchEntityException;
+use Magento\GraphQl\Quote\GetMaskedQuoteIdByReservedOrderId;
+use Magento\TestFramework\Helper\Bootstrap;
+use Magento\TestFramework\TestCase\GraphQlAbstract;
+
+class GiftMessageTest extends GraphQlAbstract
```
Comment (danielrenaud) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/GiftMessage/Cart/GiftMessageTest.php:16]: Could you add a negative test case? Where you attempt to retrieve gift message for a cart that has none.

**Review by danielrenaud — CHANGES_REQUESTED** [review]

File: app/code/Magento/GiftMessageGraphQl/composer.json (hunk 6 above)
Comment (Usik2203) [reply app/code/Magento/GiftMessageGraphQl/composer.json]: **Done**

File: app/code/Magento/GiftMessageGraphQl/Model/Resolver/Cart/GiftMessage.php (hunk 7 above)
Comment (Usik2203) [reply app/code/Magento/GiftMessageGraphQl/Model/Resolver/Cart/GiftMessage.php]: **Done**

File: dev/tests/api-functional/testsuite/Magento/GraphQl/GiftMessage/Cart/GiftMessageTest.php (hunk 8 above)
Comment (Usik2203) [reply dev/tests/api-functional/testsuite/Magento/GraphQl/GiftMessage/Cart/GiftMessageTest.php:16]: **Done**

**Comment by rogyar** [comment]
Hi @Usik2203. Everything is good from my end. Could you check the failing static tests, please?

**Comment by Usik2203** [comment]
@magento run all tests

File: app/code/Magento/GiftMessageGraphQl/Model/Resolver/Cart/GiftMessage.php [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        }
+        $cart = $value['model'];
+
+        try {
+            $giftCartMessage = $this->cartRepository->get($cart->getId());
+        } catch (LocalizedException $e) {
+            throw new GraphQlInputException(__('Can\'t load cart.'));
+        }
+
+        return [
```
Comment (danielrenaud) [inline app/code/Magento/GiftMessageGraphQl/Model/Resolver/Cart/GiftMessage.php:88]: I believe it is better to return null if the cart does not have a gift message, rather than returning an object with empty fields. This would align more with schema as you have specified theses fields as non-nullable.

File: dev/tests/api-functional/testsuite/Magento/GraphQl/GiftMessage/Cart/GiftMessageTest.php [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        gift_message {
+            to
+            from
+            message
+        }
+    }
+}
+QUERY;
+        $response = $this->graphQlQuery($query);
+        self::assertArrayHasKey('gift_message', $response['cart']);
```
Comment (danielrenaud) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/GiftMessage/Cart/GiftMessageTest.php]: Can we also assert that the values are correct, and not just existing

**Comment by Usik2203** [comment]
@magento run all tests

**Comment by Usik2203** [comment]
@magento run all tests

File: app/code/Magento/GiftMessageGraphQl/Model/Resolver/Cart/GiftMessage.php (hunk 9 above)
Comment (Usik2203) [reply app/code/Magento/GiftMessageGraphQl/Model/Resolver/Cart/GiftMessage.php:88]: **Done**

File: dev/tests/api-functional/testsuite/Magento/GraphQl/GiftMessage/Cart/GiftMessageTest.php (hunk 10 above)
Comment (Usik2203) [reply dev/tests/api-functional/testsuite/Magento/GraphQl/GiftMessage/Cart/GiftMessageTest.php]: **Done**

**Comment by Usik2203** [comment]
@magento run all tests

**Comment by eduard13** [comment]
@magento run all tests

**Comment by Usik2203** [comment]
@magento run all tests

**Review by danielrenaud — APPROVED** [review]

**Comment by Usik2203** [comment]
@magento run all tests

**Comment by Usik2203** [comment]
@magento run all tests

**Comment by Usik2203** [comment]
@magento run all tests

**Comment by eduard13** [comment]
@magento run Static Tests

**Comment by Usik2203** [comment]
@magento run Static Tests

**Comment by Usik2203** [comment]
@magento run Static Tests

**Comment by Usik2203** [comment]
@magento run Static Tests

**Comment by Usik2203** [comment]
@magento run Magento Health Index

**Comment by Usik2203** [comment]
@magento run Static Tests

**Comment by Usik2203** [comment]
@magento run Static Tests

**Review by rogyar — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @rogyar, thank you for the review.
[ENGCOM-7587](https://jira.corp.magento.com/browse/ENGCOM-7587) has been created to process this Pull Request

**Comment by paliarush** [comment]
Verified that no new `@api` or interfaces were introduced in scope of this PR, so approval by the architecture team is not required.

**Comment by engcom-Bravo** [comment]
:heavy_check_mark: QA Passed

The GraphQL query
```graphql
query {
  cart(
    cart_id: "8T3x445Ys2Scga6yWHAGVP9GYUPR8ua2"
  ) {
    gift_message {
      from
      message
      to
    }
  }
}
```
returns
```graphql
{
  "data": {
    "cart": {
      "gift_message": {
        "from": "From",
        "message": "Message",
        "to": "To"
      }
    }
  }
}
```
